kingpin/actors/packagecloud.py

```diff
@@ -0,0 +1,486 @@
+
+from kingpin.actors import base
+from kingpin.actors import exceptions
+from kingpin.constants import REQUIRED
+
+log = logging.getLogger(__name__)
+
+__author__ = 'Charles McLaughlin <charles@nextdoor.com>'
+
+ACCOUNT = os.getenv('PACKAGECLOUD_ACCOUNT', None)
+TOKEN = os.getenv('PACKAGECLOUD_TOKEN', None)
+
+
+class PackagecloudAPI(api.RestConsumer):
+
+    ENDPOINT = 'https://packagecloud.io/api/v1/'
+    CONFIG = {
+        'attrs': {
+            'packages': {
+                'path': ('repos/%account%/%repo%/packages.json'
+                         '?per_page={}'.format(sys.maxint)),
+                'http_methods': {'get': {}}
+            },
+            'delete': {
+                'path': 'repos/%account%/%repo%/%distro_version%/%filename%',
+                'http_methods': {'delete': {}}
+            },
+        },
+        'auth': {
+            'user': TOKEN,
+            'pass': ''
+        }
+    }
+
+
+class PackagecloudBase(base.BaseActor):
+
+    """Simple packagecloud Abstract Base Object"""
+
+    def __init__(self, *args, **kwargs):
+        """Check required environment variables."""
+        super(PackagecloudBase, self).__init__(*args, **kwargs)
+
+        if not ACCOUNT:
+            raise exceptions.InvalidCredentials(
+                'Missing the "PACKAGECLOUD_ACCOUNT" environment variable.')
+
+        if not TOKEN:
+            raise exceptions.InvalidCredentials(
+                'Missing the "PACKAGECLOUD_TOKEN" environment variable.')
+
+        rest_client = api.RestClient()
+        self._packagecloud_client = PackagecloudAPI(client=rest_client)
+
+    @gen.coroutine
+    def _get_all_packages(self, repo):
+        """Simple method for fetching a dictionary of all packages in a repo
+
+        Args:
+            repo: name of the packagecloud repo to fetch from
+
+        Returns:
+            A hash of the packages.
+        """
+        all_packages = yield self._packagecloud_client.packages(
+            token=TOKEN, account=ACCOUNT, repo=repo).http_get()
+        raise gen.Return(all_packages)
+
+    def _get_package_versions(self, name, all_packages):
+        """Find all versions of a given package.
+
+        Args:
+            name: name of the package to look for
+            all_packages: hash of all the packages, as returned by the API
+
+        Returns:
+            A hash of package versions sorted by creation date
+        """
+        versions = [{
+            'created_at': datetime.datetime.strptime(
+                package['created_at'], '%Y-%m-%dT%H:%M:%S.%fZ'),
+            'distro_version': package['distro_version'],
+            'filename': package['package_html_url'].split('/')[-1],
+            'name': package['name']
+        } for package in all_packages if package['name'] == name]
+
+        versions.sort(key=lambda x: x.get('created_at'), reverse=False)
+        return versions
+
+    def _get_packages_list_to_delete(self, packages_to_delete, all_packages):
```

Comment:
In fact, this code doesn't deal with any deletion at all. I'd change the method name to something like `_filter_packages(self, packages, regex)`